File: tests/language-support/java/src/bookshelf/Storage.java
package bookshelf;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class Storage {

    public static void addBook(Shelf shelf, Book book) {
        if (isFull(shelf)) {
            throw new RuntimeException(
                "Shelf '" + shelf.name + "' is at capacity");
        }
        shelf.books.add(book);
    }

    public static Book removeBook(Shelf shelf, String isbn) {
        for (int i = 0; i < shelf.books.size(); i++) {
            if (shelf.books.get(i).isbn.equals(isbn)) {
                return shelf.books.remove(i);
            }
        }
        throw new RuntimeException(
            "No book with ISBN '" + isbn + "' found");
    }

    public static List<Book> findByAuthor(Shelf shelf, String author) {
        List<Book> results = new ArrayList<>();
        String query = author.toLowerCase();
        for (Book book : shelf.books) {
            if (book.author.toLowerCase().contains(query)) {
                results.add(book);
            }
        }
        return results;
    }

    public static List<Book> findByYearRange(Shelf shelf,
                                             int startYear, int endYear) {
        List<Book> results = new ArrayList<>();
        for (Book book : shelf.books) {
            if (book.year >= startYear && book.year <= endYear) {
                results.add(book);
            }
        }
        return results;
    }

    public static List<Book> sortByTitle(Shelf shelf) {
        List<Book> sorted = new ArrayList<>(shelf.books);
        sorted.sort((a, b) -> a.title.compareTo(b.title));
        return sorted;
    }

    public static boolean isFull(Shelf shelf) {
        return shelf.books.size() >= shelf.capacity;
    }

    public static String generateReport(Shelf shelf) {
        int total = shelf.books.size();
        int availableCount = 0;
        for (Book b : shelf.books) {
            if (b.available) availableCount++;
        }
        int availPct = total > 0 ? (availableCount * 100 / total) : 0;
        int capPct = shelf.capacity > 0
            ? (total * 100 / shelf.capacity) : 0;

        Map<String, Integer> authorCounts = new LinkedHashMap<>();
        for (Book b : shelf.books) {
            authorCounts.merge(b.author, 1, Integer::sum);
        }

        int minYear = shelf.books.get(0).year;
        int maxYear = shelf.books.get(0).year;
        for (Book b : shelf.books) {
            if (b.year < minYear) minYear = b.year;
            if (b.year > maxYear) maxYear = b.year;
        }

        StringBuilder sb = new StringBuilder();
        sb.append("=== Library Report: ").append(shelf.name)
          .append(" ===\n");
        sb.append("Total books: ").append(total).append("\n");
        sb.append("Available: ").append(availableCount).append(" / ")
          .append(total).append(" (").append(availPct).append("%)\n");
        sb.append("Capacity: ").append(total).append(" / ")
          .append(shelf.capacity).append(" (").append(capPct)
          .append("% full)\n");
        sb.append("\n");
        sb.append("Authors (").append(authorCounts.size())
          .append(" unique):\n");
        for (Map.Entry<String, Integer> entry : authorCounts.entrySet()) {
            sb.append("  - ").append(entry.getKey()).append(" (")
              .append(entry.getValue()).append(" books)\n");
        }
        sb.append("\n");
        sb.append("Year range: ").append(minYear).append(" - ")
          .append(maxYear).append("\n");
        sb.append("\n");
        sb.append("Books by availability:\n");
        for (Book book : shelf.books) {
            char marker = book.available ? '+' : '-';
            sb.append("  [").append(marker).append("] ")
              .append(book.title).append(" by ").append(book.author)
              .append(" (").append(book.year).append(") - ")
              .append(book.isbn).append("\n");
        }
        return sb.toString();
    }
}

File: tests/language-support/clojure/src/bookshelf/storage.clj
(ns bookshelf.storage
  (:require [clojure.string :as str]))

(defn is-full
  "Check whether the shelf has reached its capacity."
  [shelf]
  (>= (count (:books shelf)) (:capacity shelf)))

(defn add-book
  "Add a book to the shelf. Returns updated shelf."
  [shelf book]
  (when (is-full shelf)
    (throw (ex-info (str "Shelf '" (:name shelf) "' is at capacity (" (:capacity shelf) ")")
                    {:shelf shelf})))
  (update shelf :books conj book))

(defn remove-book
  "Remove a book by ISBN. Returns updated shelf."
  [shelf isbn]
  (let [books (:books shelf)
        idx   (first (keep-indexed (fn [i b] (when (= (:isbn b) isbn) i)) books))]
    (when (nil? idx)
      (throw (ex-info (str "No book with ISBN '" isbn "' found on shelf") {:isbn isbn})))
    (assoc shelf :books (into (subvec books 0 idx) (subvec books (inc idx))))))

(defn find-by-author
  "Find books by author using case-insensitive substring match."
  [shelf author]
  (let [query (str/lower-case author)]
    (filterv #(str/includes? (str/lower-case (:author %)) query)
             (:books shelf))))

(defn find-by-year-range
  "Find books published within an inclusive year range."
  [shelf start-year end-year]
  (filterv #(and (>= (:year %) start-year) (<= (:year %) end-year))
           (:books shelf)))

(defn sort-by-title
  "Return books sorted by title without mutating the shelf."
  [shelf]
  (sort-by :title (:books shelf)))

(defn generate-report
  "Generate a multi-line formatted report for the shelf."
  [shelf]
  (let [books           (:books shelf)
        total           (count books)
        available-count (count (filter :available books))
        avail-pct       (if (pos? total) (quot (* available-count 100) total) 0)
        cap-pct         (if (pos? (:capacity shelf)) (quot (* total 100) (:capacity shelf)) 0)
        author-order    (distinct (map :author books))
        author-counts   (frequencies (map :author books))
        unique-authors  (count author-order)
        years           (map :year books)
        min-year        (if (seq years) (apply min years) 0)
        max-year        (if (seq years) (apply max years) 0)]
    (str/join
     "\n"
     (concat
      [(str "=== Library Report: " (:name shelf) " ===")
       (str "Total books: " total)
       (str "Available: " available-count " / " total " (" avail-pct "%)")
       (str "Capacity: " total " / " (:capacity shelf) " (" cap-pct "% full)")
       ""
       (str "Authors (" unique-authors " unique):")]
      (map (fn [author]
             (str "  - " author " (" (get author-counts author) " books)"))
           author-order)
      [""
       (str "Year range: " min-year " - " max-year)
       ""
       "Books by availability:"]
      (map (fn [book]
             (str "  [" (if (:available book) "+" "-") "] "
                  (:title book) " by " (:author book)
                  " (" (:year book) ") - " (:isbn book)))
           books)))))
